File: QuickBLE-Apple/iOS_macOS/QuickBLE/BLEClient.swift
import Foundation
import CoreBluetooth

#if os(iOS)
    import UIKit
#elseif os(macOS)
    import Cocoa
#endif

@objc public class BLEClient: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate{
    //MARK: Properties and Variables
    
    private let UNKNOWN_WRITING_DEV_ADDRESS = "unknown"
    
    private(set) var delegate: BLEDelegate!
    // Platform Specific Objects
    private var serviceObjects: [CBService] = []
    private var characteristicObjects: [CBCharacteristic] = []
    private var descriptorObjects: [CBDescriptor] = []
    //UUIDs for Gatt Objects
    private(set) var services: [String] = []
    private(set) var characteristics: [String] = []
    private(set) var descriptors: [String] = []
    private(set) var scanServices: [String] = []
    
    // Platform Specific Options
    var continuousScan = false
    
    // Status
    private(set) var isScanning = false;
    var isConnected: Bool{
        get{
            return connectedPeripheral != nil
        }
    }
    
    // Objective-C Mess:
    @objc public func getIsScanning() -> Bool{
        return isScanning;
    }
    @objc public func getIsConnected() -> Bool{
        return isConnected;
    }
    @objc public func getServices() -> [String]{
        return services
    }
    @objc public func getCharacteristics() -> [String]{
        return characteristics
    }
    @objc public func getDescriptors() -> [String]{
        return descriptors;
    }
    @objc public func getScanServices() -> [String]{
        return scanServices;
    }
    @objc public func getContinuousScan() -> Bool{
        return continuousScan;
    }
    @objc public func setContinuousScan(_ continuous: Bool){
        continuousScan = continuous;
    }
    @objc public func setRequestBtTitle(_ title: String){
        REQUEST_BT_TITLE = title;
    }
    @objc public func getRequestBtTitle() -> String{
        return REQUEST_BT_TITLE;
    }
    @objc public func setRequestBtMessage(_ message: String){
        REQUEST_BT_MESSAGE = message;
    }
    @objc public func getRequestBtMessage() -> String{
        return REQUEST_BT_MESSAGE;
    }
    @objc public func setRequestBtConfirm(_ confirm: String){
        REQUEST_BT_CONFIRM = confirm;
    }
    @objc public func getRequestBtConfrim() -> String{
        return REQUEST_BT_CONFIRM;
    }
    @objc public func setRequestBtDeny(_ deny: String){
        REQUEST_BT_DENY = deny;
    }
    @objc public func getRequestBtDeny() -> String{
        return REQUEST_BT_DENY;
    }
    
    
    // Keep track of detected devices
    private var deviceAddresses: [String] = []
    private(set) var devices: [CBPeripheral] = []

    // Enable BT Dialog Text
    var REQUEST_BT_TITLE = "Bluetooth Required"
    var REQUEST_BT_MESSAGE = "Please enable bluetooth in Settings."
    var REQUEST_BT_CONFIRM = "Settings"
    var REQUEST_BT_DENY = "Cancel"
    
    //Apple Specific Bluetooth Objects
    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    
    //MARK: Constructor
    @objc public init(_ delegate: BLEDelegate){
        super.init()
        self.delegate = delegate
        self.centralManager = CBCentralManager(delegate: self, queue: nil, options: [CBCentralManagerOptionShowPowerAlertKey: false])
    }
    
    /**
     Expand UUID if it is only 4 chars (stick it into BLE Base UUID)
    */
    private func expand(uuidString: String) -> String{
        if(uuidString.count == 4){
            // Add the BLE Base UUID
            return "0000" + uuidString + "-0000-1000-8000-00805F9B34FB";
        }
        return uuidString
    }
    
    //MARK: Client Control
    
    /**
     Scan for devices advertising a certain service
     - parameter service: The service to scan for
     - parameter scanFor: Whether or not to scan for the service
    */
    @objc public func scanForService(_ service: String, scanFor: Bool = true){
        if(scanFor && !scanServices.contains(service.uppercased())){
            scanServices.append(service.uppercased())
        }else if(!scanFor && scanServices.contains(service.uppercased())){
            scanServices.remove(at: scanServices.index(of: service.uppercased())!)
        }
    }
    /**
     Check bluetooth low energy client compatibility for the device
     - returns: An error code (BtError)
    */
    @objc public func checkBluetooth() -> Int{
        return centralManager.state.rawValue
    }
    
#if os(iOS)
    /**
     Show adialog requesting that the user enable bluetooth
     - parameter vc: The current UIViewController
     */
    @objc public func requestEnableBt(){
        let vc = UIApplication.shared.keyWindow?.rootViewController;
        let alert = UIAlertController(title: REQUEST_BT_TITLE, message: REQUEST_BT_MESSAGE, preferredStyle: .alert);
        let cancelAction = UIAlertAction(title: REQUEST_BT_DENY, style: .cancel, handler: {(action) -> Void in
            alert.dismiss(animated: true, completion: nil)
            self.delegate.onBluetoothRequestResult(false)
        })
        let settingsAction = UIAlertAction(title: REQUEST_BT_CONFIRM, style: .default, handler: {(action) -> Void in
            let btSettings = URL(string: "App-Prefs:root=Bluetooth")
            if(btSettings != nil){
                if UIApplication.shared.canOpenURL(btSettings!) {
                    if #available(iOS 10.0, *) {
                        UIApplication.shared.open(btSettings!, options: [:], completionHandler: nil)
                    } else {
                        UIApplication.shared.openURL(btSettings!)
                    }
                }
            }
            alert.dismiss(animated: true, completion: nil)
            self.delegate.onBluetoothRequestResult(true)
        })
        alert.addAction(settingsAction)
        alert.addAction(cancelAction)
        vc?.present(alert, animated: true, completion: nil);
    }
#elseif os(macOS)
    /**
     Show adialog requesting that the user enable bluetooth
     */
    @objc public func requestEnableBt(){
        let alert = NSAlert();
        alert.addButton(withTitle: REQUEST_BT_CONFIRM)
        alert.addButton(withTitle: REQUEST_BT_DENY)
        alert.messageText = REQUEST_BT_TITLE
        alert.informativeText = REQUEST_BT_MESSAGE
        alert.alertStyle = .warning
        if(alert.runModal() == NSApplication.ModalResponse.alertFirstButtonReturn){
            self.delegate.onBluetoothRequestResult(true)
            // Yes this is absolutely rediculous, even by apple's standards
            // To open bt preferences: run command in bash
            let task = Process()
            task.launchPath = "/usr/bin/open"
            task.arguments = ["/System/Library/PreferencePanes/Bluetooth.prefPane"]
            task.launch()
        }else{
            self.delegate.onBluetoothRequestResult(false)
        }
    }
#endif
    /**
     Start scanning for devices
     - returns: An error code (BtError)
    */
    @objc public func scanForDevices() -> Int{
        if(!isScanning){
            services.removeAll()
            characteristics.removeAll()
            descriptors.removeAll()
            serviceObjects.removeAll()
            characteristicObjects.removeAll()
            descriptorObjects.removeAll()
            let error = checkBluetooth()
            if(error != BtError.None){
                return error
            }
            var uuids: [CBUUID] = []
            for s in scanServices{
                uuids.append(CBUUID(string: s))
            }
            centralManager.scanForPeripherals(withServices: uuids, options: [CBCentralManagerScanOptionAllowDuplicatesKey: continuousScan])
            isScanning = true;
            return BtError.None
        }else{
            return BtError.AlreadyRunning
        }
    }
    /**
     Stop scanning
    */
    @objc public func stopScanning(){
        if(isScanning){
            centralManager.stopScan()
            isScanning = false;
        }
    }
    
    /**
     Connect to a specified device
     - parameter deviceAddress: The address of the device to connect to
    */
    @objc public func connectToDevice(_ deviceAddress: String){
        if(connectedPeripheral != nil){
            disconnect()
        }
        let dev = devices.first(where: {$0.identifier.uuidString.uppercased() == deviceAddress.uppercased()})
        if(dev == nil){
            return;
        }
        centralManager.connect(dev!, options: [CBConnectPeripheralOptionNotifyOnConnectionKey: true, CBConnectPeripheralOptionNotifyOnDisconnectionKey: true, CBConnectPeripheralOptionNotifyOnNotificationKey: true])
    }
    /**
     Disconnect from server if connected
    */
    @objc public func disconnect(){
        if(isConnected){
            centralManager.cancelPeripheralConnection(connectedPeripheral!)
            services.removeAll()
            characteristics.removeAll()
            descriptors.removeAll()
            serviceObjects.removeAll()
            characteristicObjects.removeAll()
            descriptorObjects.removeAll()
            connectedPeripheral = nil
        }
    }
    /**
     Subscribe to a characteristic to recieve notifications when it's value is changed
     - parameter characteristic: The characteristic to subscribe to
     - parameter subscribe: Whether or not to subscribe to the characteristic (false to unsubscribe)
    */
    @objc public func subscribeToCharacteristic(_ characteristic: String, subscribe: Bool = true){
        guard let char = getCharacteristic(uuid: CBUUID(string: characteristic)) else{
            return
        }
        if(connectedPeripheral != nil){
            connectedPeripheral?.setNotifyValue(subscribe, for: char)
        }
    }
    
    //MARK Characteristics and Descriptors
    
    /**
     Read a value from a characteristic
     - parameter characteristic: The characteristic to read
    */
    @objc public func readCharacteristic(_ characteristic: String){
        guard let char = getCharacteristic(uuid: CBUUID(string: characteristic)) else{
            return
        }
        if(connectedPeripheral != nil){
            connectedPeripheral!.readValue(for: char)
        }
    }
    /**
     Write a value to a characteristic
     - parameter characteristic: The characteristic to write
     - parameter data: The value to write (as Data)
    */
    @objc public func writeCharacteristic(_ characteristic: String, data: Data?){
        guard let char = self.getCharacteristic(uuid: CBUUID(string: characteristic)) else{
            return
        }
        if self.connectedPeripheral != nil, data != nil{
            self.connectedPeripheral!.writeValue(data!, for: char, type: .withResponse)
        }
    }
    /**
     Read a value from a descriptor
     - parameter descriptor: The descriptor to read
    */
    @objc public func readDescriptor(_ descriptor: String){
        guard let desc = getDescriptor(uuid: CBUUID(string: descriptor)) else{
            return
        }
        if(connectedPeripheral != nil){
            connectedPeripheral!.readValue(for: desc)
        }
    }
    /**
     Write a value to a descriptor
     - parameter descriptor: The descriptor to write
     - parameter data: The value to write (as Data)
    */
    @objc public func writeDescriptor(_ descriptor: String, data: Data?){
        guard let desc = getDescriptor(uuid: CBUUID(string: descriptor)) else{
            return
        }
        if connectedPeripheral != nil, data != nil{
            connectedPeripheral!.writeValue(data!, for: desc)
        }
    }
    
    /**
     Get a service from a UUID
     -parameter uuid: The uuid of the service
     - returns: The CBService or nil
    */
    private func getService(uuid: CBUUID) -> CBService?{
        return serviceObjects.first(where: { $0.uuid.isEqual(uuid) })
    }
    /**
     Get a characteristic from a UUID
     -parameter uuid: The uuid of the characteristic
     - returns: The CBCharacteristic or nil
     */
    private func getCharacteristic(uuid: CBUUID) -> CBCharacteristic?{
        return characteristicObjects.first(where: { $0.uuid.isEqual(uuid) })
    }
    /**
     Get a descriptor from a UUID
     -parameter uuid: The uuid of the descriptor
     - returns: The CBDescriptor or nil
     */
    private func getDescriptor(uuid: CBUUID) -> CBDescriptor?{
        return descriptorObjects.first(where: { $0.uuid.isEqual(uuid) })
    }
    
    /**
     Check if the client has a service
     - parameter service: The service to search for
     - returns: Whether or not the client has the service
     */
    @objc public func hasService(_ service: String) -> Bool{
        return services.contains(service.uppercased())
    }
    /**
     Check if the client has a characteristic
     - parameter characteristic: The service to search for
     - returns: Whether or not the client has the characteristic
     */
    @objc public func hasCharacteristic(_ characteristic: String) -> Bool{
        return characteristics.contains(characteristic.uppercased())
    }
    /**
     Check if the client has a descriptor
     - parameter descriptor: The descriptor to search for
     - returns: Whether or not the client has the descriptor
     */
    @objc public func hasDescriptor(_ descriptor: String) -> Bool{
        return descriptors.contains(descriptor.uppercased())
    }
    
    //MARK: CBCentralManagerDelegate
    @objc public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        let result = central.state.rawValue == BtError.None;
        if (!result){
            stopScanning()
            disconnect()
        }
        delegate.onBluetoothPowerChange(result)
    }
    @objc public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        peripheral.delegate = self
        peripheral.discoverServices(nil)
        connectedPeripheral = peripheral
        services.removeAll()
        characteristics.removeAll()
        descriptors.removeAll()
        delegate.onConnectToDevice(address: peripheral.identifier.uuidString.uppercased(), name: peripheral.name, success: true)
    }
    @objc public func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        delegate.onConnectToDevice(address: peripheral.identifier.uuidString.uppercased(), name: peripheral.name, success: false)
    }
    @objc public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        delegate.onDisconnectFromDevice(address: peripheral.identifier.uuidString.uppercased(), name: peripheral.name)
    }
    @objc public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        if(!deviceAddresses.contains(peripheral.identifier.uuidString)){
            devices.append(peripheral)
            deviceAddresses.append(peripheral.identifier.uuidString)
        }
        delegate.onDeviceDiscovered(address: peripheral.identifier.uuidString.uppercased(), name: peripheral.name, rssi: RSSI.intValue)
    }
    
    // Keep track of how many responses there should be and have been for discovering chars/incServices/descs
    // This is Apple Specific
    private var charCallCount = 0
    private var incCallCount = 0
    private var descCallCount = 0
    
    private var completedChar = 0
    private var completedInc = 0
    private var completedDesc = 0
    
    //MARK: CBPeripheralDelegate
    @objc public func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        peripheral.delegate = self
        if(peripheral.services != nil){
            serviceObjects.append(contentsOf: peripheral.services!)
            charCallCount = serviceObjects.count
            incCallCount = serviceObjects.count
            descCallCount = 0
            for service in peripheral.services!{
                services.append(expand(uuidString: service.uuid.uuidString))
                peripheral.discoverIncludedServices(nil, for: service)
                peripheral.discoverCharacteristics(nil, for: service);
            }
        }
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didDiscoverIncludedServicesFor service: CBService, error: Error?) {
        completedInc += 1
        peripheral.delegate = self
        if(service.includedServices != nil){
            serviceObjects.append(contentsOf: service.includedServices!)
            charCallCount += service.includedServices!.count
            for s in service.includedServices!{
                services.append(expand(uuidString: s.uuid.uuidString))
                peripheral.discoverCharacteristics(nil, for: s)
            }
        }
        allDiscovered()
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        completedChar += 1
        peripheral.delegate = self
        if(service.characteristics != nil){
            characteristicObjects.append(contentsOf: service.characteristics!)
            descCallCount += service.characteristics!.count
            for characteristic in service.characteristics!{
                characteristics.append(expand(uuidString: characteristic.uuid.uuidString))
                peripheral.discoverDescriptors(for: characteristic)
            }
        }
        allDiscovered()
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didDiscoverDescriptorsFor characteristic: CBCharacteristic, error: Error?) {
        completedDesc += 1
        if(characteristic.descriptors != nil){
            descriptorObjects.append(contentsOf: characteristic.descriptors!)
            for descriptor in characteristic.descriptors!{
                descriptors.append(expand(uuidString: descriptor.uuid.uuidString))
            }
        }
        allDiscovered()
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        delegate.onCharacteristicRead(expand(uuidString: characteristic.uuid.uuidString).uppercased(), writingDeviceAddress: UNKNOWN_WRITING_DEV_ADDRESS, success: error == nil, value: characteristic.value)
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
        delegate.onCharacteristicWrite(expand(uuidString: characteristic.uuid.uuidString).uppercased(), success: error == nil, value: characteristic.value)
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor descriptor: CBDescriptor, error: Error?) {
        delegate.onDescriptorRead(expand(uuidString: descriptor.uuid.uuidString).uppercased(), writingDeviceAddress: UNKNOWN_WRITING_DEV_ADDRESS, success: error == nil, value: descriptor.value as? Data)
    }
    @objc public func peripheral(_ peripheral: CBPeripheral, didWriteValueFor descriptor: CBDescriptor, error: Error?) {
        delegate.onDescriptorWrite(expand(uuidString: descriptor.uuid.uuidString).uppercased(), success: error == nil, value: descriptor.value as? Data)
    }
    
    // Check if all services chars and descs have been discovered
    // All chars must be discovered before the client can subscribe to notifications (this is solution on iOS and macOS)
    private func allDiscovered(){
        if charCallCount <= completedChar, incCallCount <= completedInc, descCallCount <= completedDesc{
            delegate.onServicesDiscovered()
        }
    }
}

File: QuickBLE-Android/app/src/main/java/com/eyeballmonkeymedia/quickble/BLEClient.kt
package com.mb3hel.quickble

import android.app.AlertDialog
import android.bluetooth.*
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.media.VolumeShaper
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import android.support.annotation.RequiresApi
import android.util.Log
import java.util.*
import kotlin.collections.ArrayList
import kotlin.concurrent.thread

@RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR2)
class BLEClient (val context: Context, val delegate: BLEDelegate, val useNewMethod: Boolean = true){

    //region Variables and Properties

    val gattOperationQueue = GattOperationQueue(this)

    private val UNKNOWN_WRITING_DEVICE_ADDRESS = "unknown";

    // Platform Specific Objects
    private var serviceObjects = ArrayList<BluetoothGattService>()
    private var characteristicObjects = ArrayList<BluetoothGattCharacteristic>()
    private var descriptorObjects = ArrayList<BluetoothGattDescriptor>()
    // UUIDs for Gatt Objects
    /**
     * The services available on the connected server
     */
    var services = ArrayList<String>()
        private set
    /**
     * The characteristics available on the connected server
     */
    var characteristics = ArrayList<String>()
        private set
    /**
     * The descriptors available on the connected server
     */
    var descriptors = ArrayList<String>()
        private set
    /**
     * The services that are being scanned for when scanning for devices
     */
    var scanServices = ArrayList<String>()
        private set

    // Platform Specific Options
    var scanMode = if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) ScanMode.Balanced else ScanMode.NotApplicable
        set(value) {
            if(useNewMethod){
                field = value
            }
        }

    // Status
    /**
     * Is the client scanning for servers
     */
    var isScanning = false
        private set
    /**
     * Is the client connected to a server
     */
    var isConnected = false
        get() = (gattConnection != null)
        private set

    // Keep track of detected devices
    private var deviceAddresses = ArrayList<String>()
    var devices = ArrayList<BluetoothDevice>()
        private set

    // Android Specific Bluetooth Objects
    private val mainThread = Handler(Looper.getMainLooper())
    private val btManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val btAdapter = btManager.adapter
    var gattConnection: BluetoothGatt? = null
        private set

    // Enable BT Dialog Text
    var REQUEST_BT_TITLE = "Bluetooth Required"
    var REQUEST_BT_MESSAGE = "Enable bluetooth?"
    var REQUEST_BT_CONFIRM = "Yes"
    var REQUEST_BT_DENY = "No"
    //endregion

    /**
     * Create a new QuickBLE Client for Gatt central role
     */
    init{
        // Watch for Bluetooth Power Changes
        thread(start = true){
            var lastState = false
            while(true){
                val state = btAdapter.isEnabled
                if(state != lastState){
                    lastState = state
                    if(!state){
                        stopScanning()
                        disconnect()
                    }
                    mainThread.post {
                        delegate.onBluetoothPowerChanged(state)
                    }
                }
                Thread.sleep(100)
            }
        }
    }

    //region Client Control
    /**
     * Scan for devices advertising a certain service (only available when using new scan method on Lolipop+ devices)
     * @param service The service to scan for
     * @param scanFor Whether or not to scan for the service
     */
    fun scanForService(service: String, scanFor: Boolean = true){
        if(scanFor && !scanServices.contains(service.toUpperCase())){
            scanServices.add(service.toUpperCase())
        }else if(!scanFor && scanServices.contains(service.toUpperCase())){
            scanServices.remove(service.toUpperCase())
        }
    }
    /**
     * Check bluetooth low energy client compatibility for the device
     * @return An error code (BtError)
     */
    fun checkBluetooth(): Int{
        if (!context.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH))
            return BtError.NoBluetooth
        if (!context.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE))
            return BtError.NoBLE
        if (btAdapter == null || !btAdapter.isEnabled)
            return BtError.Disabled
        return BtError.None
    }
    /**
     * Show a dialog requesting that the user enable bluetooth
     */
    fun requestEnableBt(){
        val builder = AlertDialog.Builder(context, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
        builder.setTitle(REQUEST_BT_TITLE).setMessage(REQUEST_BT_MESSAGE)
        builder.setPositiveButton(REQUEST_BT_CONFIRM, { _, _ ->
            btAdapter.enable()
            delegate.onBluetoothRequestResult(true)
        })
        builder.setNegativeButton(REQUEST_BT_DENY, {_, _  ->
            delegate.onBluetoothRequestResult(false)
        })
        builder.create().show()
    }

    /**
     * Start scanning for devices
     * @return An error code (BtError)
     */
    fun scanForDevices(): Int{
        if(!isScanning && !isConnected){
            services.clear()
            characteristics.clear()
            descriptors.clear()
            serviceObjects.clear()
            characteristicObjects.clear()
            descriptorObjects.clear()
            val error = checkBluetooth()
            if(error != BtError.None){
                return error
            }
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && useNewMethod){
                val settings = ScanSettings.Builder().setScanMode(scanMode).build()
                val filters = ArrayList<ScanFilter>()
                scanServices.forEach {
                    filters.add(ScanFilter.Builder().setServiceUuid(ParcelUuid(UUID.fromString(it))).build())
                }
                btAdapter.bluetoothLeScanner.startScan(filters, settings, scanCallback)

            }else{
                if(scanServices.size > 0){
                    val uuids = ArrayList<UUID>()
                    scanServices.forEach {
                        uuids.add(UUID.fromString(it))
                    }
                    btAdapter.startLeScan(uuids.toTypedArray(), leScanCallback)
                }else {
                    btAdapter.startLeScan(leScanCallback)
                }
            }
            isScanning = true
            return BtError.None
        }else{
            return BtError.AlreadyRunning
        }
    }
    /**
     * Stop scanning
     */
    fun stopScanning(){
        if(isScanning){
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && useNewMethod){
                btAdapter.bluetoothLeScanner.stopScan(scanCallback)
            }else{
                btAdapter.stopLeScan(leScanCallback)
            }
            isScanning = false
        }
    }

    /**
     * Connect to a specified device
     * @param deviceAddress The address of the device to connect to
     */
    fun connectToDevice(deviceAddress: String){
        if(gattConnection != null){
            disconnect()
        }
        gattConnection = devices.firstOrNull { it.address.toUpperCase().equals(deviceAddress.toUpperCase()) }?.connectGatt(context, false, bluetoothGattCallback)
        gattConnection?.connect()
    }
    /**
     * Disconnect from server if connected
     */
    fun disconnect(){
        if(isConnected){
            services.clear()
            characteristics.clear()
            descriptors.clear()
            serviceObjects.clear()
            characteristicObjects.clear()
            descriptorObjects.clear()
            gattConnection?.disconnect()
            gattConnection = null
            isConnected = false
        }
    }

    /**
     * Subscribe to a characteristic to receive notifications when is's value is changed
     * @param characteristic The characteristic to subscribe to
     * @param subscribe Whether not to subscribe to the characteristic (false to unsubscribe)
     */
    fun subscribeToCharacteristic(characteristic: String, subscribe: Boolean = true) {
        val char = getCharacteristic(UUID.fromString(characteristic))
        if(char != null){
            gattConnection?.setCharacteristicNotification(char, subscribe)
            val descriptor = char.getDescriptor(UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")) // Client characteristic config UUID
            if(descriptor != null){
                descriptor.value = if(subscribe) BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE else BluetoothGattDescriptor.DISABLE_NOTIFICATION_VALUE
                gattOperationQueue.queueOperation(DelayedGattOperation(OperandType.Descriptor, OperationType.Write, descriptor, descriptor.value))
            }
        }
    }

    /**
     * Add a service (and it's included services, characteristics, and descriptors) to the lists
     * @param service The service to add
     */
    private fun addService(service: BluetoothGattService){
        if(!serviceObjects.contains(service)){
            if(service.type == BluetoothGattService.SERVICE_TYPE_PRIMARY){
                service.includedServices.asSequence().filterNot { serviceObjects.contains(it) }.forEach {
                    addService(it)
                }
            }
            service.characteristics.asSequence().filterNot { characteristicObjects.contains(it) }.forEach {
                it.descriptors.asSequence().filterNot { descriptorObjects.contains(it) }.forEach {
                    descriptorObjects.add(it)
                    descriptors.add(it.uuid.toString().toUpperCase())
                }
                characteristicObjects.add(it)
                characteristics.add(it.uuid.toString().toUpperCase())
            }
            serviceObjects.add(service)
            services.add(service.uuid.toString().toUpperCase())
        }
    }
    //endregion

    //region Characteristics and Descriptors
    /**
     * Read a value from a characteristic
     * @param characteristic The characteristic to read
     */
    fun readCharacteristic(characteristic: String){
        val char = getCharacteristic(UUID.fromString(characteristic))
        if(char != null){
            gattOperationQueue.queueOperation(DelayedGattOperation(OperandType.Characteristic, OperationType.Read, char, null))
        }else{
            mainThread.post {
                delegate.onCharacteristicRead(characteristic.toUpperCase(), "", false, null)
            }
        }
    }
    /**
     * Write a value to a characteristic
     * @param characteristic The characteristic to write
     * @param data The value to write (as bytes)
     */
    fun writeCharacteristic(characteristic: String, data: ByteArray?){
        val char = getCharacteristic(UUID.fromString(characteristic))
        if(char != null){
            gattOperationQueue.queueOperation(DelayedGattOperation(OperandType.Characteristic, OperationType.Write, char, data))
        }else{
            mainThread.post {
                delegate.onCharacteristicWrite(characteristic.toUpperCase(), false, null)
            }
        }
    }
    /**
     * Read a value from a descriptor
     * @param descriptor The descriptor to read
     */
    fun readDescriptor(descriptor: String){
        val desc = getDescriptor(UUID.fromString(descriptor))
        if(desc != null){
            gattOperationQueue.queueOperation(DelayedGattOperation(OperandType.Descriptor, OperationType.Read, desc, null))
        }else{
            mainThread.post {
                delegate.onDescriptorRead(descriptor.toUpperCase(), UNKNOWN_WRITING_DEVICE_ADDRESS, false, null)
            }
        }
    }
    /**
     * Write a value to a descriptor
     * @param descriptor The descriptor to write
     * @param data The value to write (as bytes)
     */
    fun writeDescriptor(descriptor: String, data: ByteArray?){
        val desc = getDescriptor(UUID.fromString(descriptor))
        if(desc != null){
            gattOperationQueue.queueOperation(DelayedGattOperation(OperandType.Descriptor, OperationType.Write, desc, data))
        }else{
            mainThread.post {
                delegate.onDescriptorWrite(descriptor.toUpperCase(), false, null)
            }
        }
    }

    fun handleGattOperationComplete(operation: DelayedGattOperation, success: Boolean){
        if(operation.operandType == OperandType.Characteristic){
            // Characteristic
            if(operation.operationType == OperationType.Read){
                // Read
                mainThread.post {
                    delegate.onCharacteristicRead((operation.operand as BluetoothGattCharacteristic).uuid.toString().toUpperCase(), UNKNOWN_WRITING_DEVICE_ADDRESS, success, operation.data)
                }
            }else{
                // Write
                mainThread.post {
                    delegate.onCharacteristicWrite((operation.operand as BluetoothGattCharacteristic).uuid.toString().toUpperCase(), success, operation.data)
                }
            }
        }else{
            // Descriptor
            if(operation.operationType == OperationType.Read){
                // Read
                mainThread.post {
                    delegate.onDescriptorRead((operation.operand as BluetoothGattDescriptor).uuid.toString().toUpperCase(), UNKNOWN_WRITING_DEVICE_ADDRESS, success, operation.data)
                }
            }else{
                // Write
                mainThread.post {
                    delegate.onDescriptorWrite((operation.operand as BluetoothGattDescriptor).uuid.toString().toUpperCase(), success, operation.data)
                }
            }
        }
    }

    /**
     * Get a descriptor object from a UUID
     * @return The BluetoothGattDescriptor or null
     */
    private fun getDescriptor(uuid: UUID): BluetoothGattDescriptor?{
        return descriptorObjects.asSequence().firstOrNull { it.uuid == uuid }
    }
    /**
     * Get a characteristic object from a UUID
     * @return The BluetoothGattCharacteristic or null
     */
    private fun getCharacteristic(uuid: UUID): BluetoothGattCharacteristic?{
        return characteristicObjects.asSequence().firstOrNull { it.uuid == uuid }
    }
    /**
     * Get a service object from a UUID
     * @return The BluetoothGattService or null
     */
    private fun getService(uuid: UUID): BluetoothGattService?{
        return serviceObjects.asSequence().firstOrNull { it.uuid == uuid }
    }

    /**
     * Check if the server has a service
     * @param service The service to search for
     * @return Whether or not the server has the service
     */
    fun hasService(service: String): Boolean{
        return services.contains(service.toUpperCase());
    }
    /**
     * Check if the server has a characteristic
     * @param characteristic The characteristic to search for
     * @return Whether or not the server has the characteristic
     */
    fun hasCharacteristic(characteristic: String): Boolean{
        return characteristics.contains(characteristic.toUpperCase());
    }
    /**
     * Check if the server has a descriptor
     * @param descriptor The descriptor to search for
     * @return Whether or not the server has the descriptor
     */
    fun hasDescriptor(descriptor: String): Boolean{
        return descriptors.contains(descriptor.toUpperCase());
    }

    //endregion

    //region LeScanCallback (Old Method)
    private val leScanCallback = BluetoothAdapter.LeScanCallback { device, rssi, scanRecord ->
        if(device != null){
            if(!deviceAddresses.contains(device.address)) {
                devices.add(device)
                deviceAddresses.add(device.address)
            }
            mainThread.post {
                delegate.onDeviceDiscovered(device.address.toUpperCase(), device.name, rssi)
            }
        }
    }
    //endregion

    //region ScanCallback (New Method)
    private val scanCallback = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
        object:ScanCallback(){
            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
            override fun onScanFailed(errorCode: Int) {
                super.onScanFailed(errorCode)
            }
            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
            override fun onScanResult(callbackType: Int, result: ScanResult?) {
                super.onScanResult(callbackType, result)
                if(result != null){
                    if(!deviceAddresses.contains(result.device.address)) {
                        devices.add(result.device)
                        deviceAddresses.add(result.device.address)
                    }
                    mainThread.post {
                        delegate.onDeviceDiscovered(result.device.address.toUpperCase(), result.device.name, result.rssi)
                    }
                }
            }
            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
            override fun onBatchScanResults(results: MutableList<ScanResult>?) {
                super.onBatchScanResults(results)
                if(results != null){
                    results.asSequence().forEach {
                        devices.add(it.device)
                        mainThread.post {
                            delegate.onDeviceDiscovered(it.device.address.toUpperCase(), it.device.name, it.rssi)
                        }
                    }
                }
            }
        }
    else null
    //endregion

    //region BluetoothGattCallback
    private val bluetoothGattCallback = object:BluetoothGattCallback(){
        override fun onCharacteristicRead(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
            super.onCharacteristicRead(gatt, characteristic, status)
            gattOperationQueue.onCharacteristicRead(characteristic, status)
        }
        override fun onCharacteristicWrite(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
            super.onCharacteristicWrite(gatt, characteristic, status)
            gattOperationQueue.onCharacteristicWrite(characteristic, status)
        }
        override fun onServicesDiscovered(gatt: BluetoothGatt?, status: Int) {
            super.onServicesDiscovered(gatt, status)
            if(gatt != null){
                gatt.services.asSequence().forEach {
                    addService(it)
                }
                mainThread.post{
                    delegate.onServicesDiscovered()
                }
            }
        }
        override fun onDescriptorWrite(gatt: BluetoothGatt?, descriptor: BluetoothGattDescriptor?, status: Int) {
            super.onDescriptorWrite(gatt, descriptor, status)
            gattOperationQueue.onDescriptorWrite(descriptor, status)
        }
        override fun onCharacteristicChanged(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?) {
            super.onCharacteristicChanged(gatt, characteristic)
            if(characteristic != null){
                val d = Arrays.copyOf(characteristic.value, characteristic.value.size)
                mainThread.post {
                    delegate.onCharacteristicRead(characteristic.uuid.toString().toUpperCase(), UNKNOWN_WRITING_DEVICE_ADDRESS, true, d)
                }
            }
        }
        override fun onDescriptorRead(gatt: BluetoothGatt?, descriptor: BluetoothGattDescriptor?, status: Int) {
            super.onDescriptorRead(gatt, descriptor, status)
            gattOperationQueue.onDescriptorRead(descriptor, status)
        }
        override fun onConnectionStateChange(gatt: BluetoothGatt?, status: Int, newState: Int) {
            super.onConnectionStateChange(gatt, status, newState)
            if(gatt != null && gatt == gattConnection){
                if(newState == BluetoothProfile.STATE_CONNECTED){
                    gatt.discoverServices()
                    mainThread.post {
                        delegate.onConnectToDevice(gatt.device.address.toUpperCase(), gatt.device.name, true)
                    }
                    isConnected = true
                }
                if(newState == BluetoothProfile.STATE_DISCONNECTED){
                    services.clear()
                    characteristics.clear()
                    descriptors.clear()
                    gattOperationQueue.clientDisconnected()
                    mainThread.post {
                        delegate.onDisconnectFromDevice(gatt.device.address.toUpperCase(), gatt.device.name)
                    }
                    isConnected = false
                }
            }
        }
    }
    //endregion

}